docs/writing-views.md

```diff
@@ -0,0 +1,301 @@
+view multiple times with the same content should send multiple
+messages), you should use POST.
+
+If you're updating existing data, use PATCH.
+
+If you're removing data, use DELETE.
+
+### Idempotency
+
+When writing a new API endpoint, with the exception of things like
+message sending, requests should be safe to repeat, without affecting
+the state of the server. This is *idempotency*.
+
+You will often return an error if a request was made more than once, to
+make debugging Zulip clients easier. This means that the response for
+repeated requests may not be the same, but the repeated requests won't
+change the server more than once or cause unwanted side effects.
+
+### Making changes to the database
+
+If the view does any modification to the database, that change is done
+in a helper function in `zerver/lib/actions.py`.
+
+For example, in [zerver/views/__init__.py](https://github.com/zulip/zulip/blob/master/zerver/views/__init__.py):
+
+```py
+@require_realm_admin
+@has_request_variables
+def update_realm(request, user_profile, name=REQ(validator=check_string, default=None), ...)):
+    # type: (HttpRequest, UserProfile, ...) -> HttpResponse
+    realm = user_profile.realm
+    data = {} # type: Dict[str, Any]
+    if name is not None and realm.name != name:
+        do_set_realm_name(realm, name)
+        data['name'] = 'updated'
+```
+
+and in [zerver/lib/actions.py](https://github.com/zulip/zulip/blob/master/zerver/lib/actions.py):
+
+```py
+def do_set_realm_name(realm, name):
+    # type: (Realm, text_type) -> None
+    realm.name = name
+    realm.save(update_fields=['name'])
+    event = dict(
+        type="realm",
+        op="update",
+        property='name',
+        value=name,
+    )
+    send_event(event, active_user_ids(realm))
+```
+
+`realm.save()` actually saves the changes to the realm to the database.
+
+### Calling from the web client
+
+You can use channel.<method> to call to the proper view from anywhere in
+the web client code. In this case, in [static/js/admin.js](https://github.com/zulip/zulip/blob/master/static/js/admin.js)
+
+```js
+var url = "/json/realm";
+var data = {
+    name: JSON.stringify(new_name),
+}
+channel.patch({
+    url: url,
+    data: data,
+    success: function (response_data) {
+        if (response_data.name !== undefined) {
+            ui.report_success(i18n.t("Name changed!"), name_status);
+        }
+        ...
+```
+
+If the user is logged in, this will handle passing in the session token
+automatically.
+
+
+### Calling from an API client
+
+Here's how you might handle calling from python:
+
+```py
+payload = {'name': new_name}
+
+# email and API key
+api_auth = ('hello@example.com', '0123456789abcdeFGHIJKLmnopQRSTUV')
+
+r = requests.patch(SERVER_URL + 'api/v1/realm',
+                   data=json.dumps(payload),
+                   auth=api_auth,
+                  )
+```
+
+## Legacy endpoints used by the web client
```

Comment:
I think it would help to mention what those endpoints are so people know if they've run into one; are they all under `/api/v1/`?